feat(react): Add Error Boundary component

AbhiPrasad opened the pull request:
Following up #2631

### Motivation

This PR implements an Error Boundary React component. See https://reactjs.org/docs/error-boundaries.html for more details.

This is an important API for us to support, as error boundaries are the defacto way to catch and deal with errors in React.

### Implementation

The implementation was inspired by the way we do error boundaries at Sentry as well as bvaughn's excellent [error boundary implementation](https://github.com/bvaughn/react-error-boundary)

The error boundary provides the following options:

```ts
type FallbackRender = (fallback: {
  error: Error | null;
  componentStack: string | null;
  resetError(): void;
}) => React.ReactNode;

export type ErrorBoundaryProps = {
  // if we should render Sentry's report dialog or not
  showDialog?: boolean;
  dialogOptions?: Sentry.ReportDialogOptions;
  // If render key changes and there is an error, the component is reset
  renderKey?: string | number;
  // We provide two ways to render a fallback component,
  // rendering just a component takes priority, but we
  // also support rendering using render props.
  fallback?: React.ReactNode | FallbackRender;
   // on error callback
  onError?(error: Error, componentStack: string): void;
  // callback onComponentDidMount
  onMount?(): void;
  // when `resetErrorBoundary` is called from `fallbackRender`
  onReset?(error: Error | null, componentStack: string | null): void;
  // callback onComponentWillUnmount
  onUnmount?(error: Error | null, componentStack: string | null): void;
};
```

We recommend using `fallbackRender` to render the fallback component, as it uses a render props approach. This passes `resetErrorBoundary` as render props, allowing the user to easily reset the error boundary.

The error boundary is available as a standalone component with

```ts
<ErrorBoundary>
  <Component />
</ErrorBoundary>
```

or as a HOC

```ts
withErrorBoundary(Component, errorBoundaryProps)
```

### Testing

I switched to `@testing-library/react`, it's pretty good (and seems to be the standard way to do tests in React now). It's also 209.5 kb minified, while enzyme is 463.2 kb, and the `yarn installs` definitely add up.

Current test coverage

File               |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-------------------|----------|----------|----------|----------|-------------------|
All files          |    91.49 |    72.73 |       90 |    90.91 |                   |
 errorboundary.tsx |    88.89 |       50 |    77.78 |     87.5 | 93,95,101,105,106 |
 profiler.tsx      |       95 |      100 |      100 |    94.59 |             53,56 |

### Future

After this gets merged in, we move on to creating a hooks based Profiler. I'm also looking to see how difficult it is to support `react-router` and similar.

HazAT commented on packages/react/src/errorboundary.tsx:
```diff
@@ -0,0 +1,109 @@
+  public state: ErrorBoundaryState = INITIAL_STATE;
+
+  public componentDidCatch(error: Error, { componentStack }: React.ErrorInfo): void {
+    Sentry.withScope(scope => {
+      scope.setExtra('componentStack', componentStack);
+      Sentry.captureException(error);
+    });
```
Try this, new API :)
```suggestion
      Sentry.captureException(error, {contexts: {componentStack}});
```

HazAT commented on packages/react/src/errorboundary.tsx:
```diff
@@ -0,0 +1,109 @@
+};
+
+type ErrorBoundaryState = {
+  error: Error | null;
+  componentStack: string | null;
```
```suggestion
  componentStack: string | null;
  error: Error | null;
```

kamilogorek commented on packages/react/src/errorboundary.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  public componentDidCatch(error: Error, { componentStack }: React.ErrorInfo): void {
+    Sentry.withScope(scope => {
+      scope.setExtra('componentStack', componentStack);
```
@HazAT should we move to setContext instead?

AbhiPrasad replied:
Changed to `Sentry.captureException(error, { contexts: { componentStack } });`

AbhiPrasad commented:
Update: Addressed review comments. Also added a `renderKey` prop, which will update the component if changed. This provides an easy way for the user to reset errors without having to rely on calling a reset function (for example, if the `
window.location` changes, the component will reset)

AbhiPrasad commented:
Hmm so I'm using `Sentry.captureException(error, { contexts: { react: { componentStack } } });`, and though an error event is created in Sentry, the react context does not show up. Testing with demo here: https://github.com/AbhiPrasad/sentry-react-demo

billyvg commented on packages/react/src/errorboundary.tsx:
```diff
@@ -0,0 +1,124 @@
+  dialogOptions?: Sentry.ReportDialogOptions;
+  fallback?: React.ReactNode;
+  renderKey?: string | number;
+  fallbackRender?(fallback: {
+    error: Error | null;
+    componentStack: string | null;
+    resetError(): void;
+  }): React.ReactNode;
```
Might be biased because of our app codebase, but I generally see `render*` used as the prefix -- it also makes sense to be consistent with the `fallback` prop. Should we overload `fallback`?

AbhiPrasad replied:
That makes sense to me, I'll overload `fallback`.

billyvg commented on packages/react/src/errorboundary.tsx:
```diff
@@ -0,0 +1,124 @@
+    if (showDialog) {
+      Sentry.showReportDialog(dialogOptions);
+    }
+    this.setState({ error, componentStack });
```
The docs recommend using https://reactjs.org/docs/react-component.html#static-getderivedstatefromerror to update state after an error is thrown, but we wouldn't be able to capture the component stack. Maybe add a comment about this?

AbhiPrasad replied:
Yes good point, I add a comment about it.

billyvg commented on packages/react/src/errorboundary.tsx:
```diff
@@ -0,0 +1,124 @@
+  public componentDidUpdate(prevProps: ErrorBoundaryProps): void {
+    const { error } = this.state;
+    const { renderKey, onReset } = this.props;
+    if (error !== null && !Object.is(renderKey, prevProps.renderKey)) {
```
Is renderKey necessary? Could we get the same behavior by simply using `key`?

AbhiPrasad replied:
Good point, people can just use `key` for this.

AbhiPrasad commented:
@billyvg we will be updating the usage of `Profiler` and `ErrorBoundary` in `getsentry/sentry` to test this before we GA the package.
